host/sgx/linux/exception.c

```diff
@@ -151,23 +154,31 @@ static void _register_signal_handlers(void)
         abort();
     }
 
-    // Unmask the signals we want to receive.
-    for (size_t i = 0; i < DEFAULT_SIGNALS_NUMBER; i++)
+    /* Unmask the default set of signals when vDSO is not enabled.
+     * Note that the check may be redundant because the kernel will
+     * filter out these signals when vDSO is used. */
+    if (!oe_sgx_is_vdso_enabled)
     {
-        int signal_number = default_signals[i];
-        sigdelset(&sig_action.sa_mask, signal_number);
-
-        // Set the signal handlers, and store the previous signal action into a
-        // global array.
-        if (sigaction(
-                signal_number,
-                &sig_action,
-                &g_previous_sigaction[signal_number]) != 0)
+        for (size_t i = 0; i < DEFAULT_SIGNALS_NUMBER; i++)
         {
-            abort();
+            int signal_number = default_signals[i];
+            sigdelset(&sig_action.sa_mask, signal_number);
+
+            /* Set the signal handlers, and store the previous signal action
+             * into a global array. */
+            if (sigaction(
+                    signal_number,
+                    &sig_action,
+                    &g_previous_sigaction[signal_number]) != 0)
+            {
+                abort();
+            }
         }
     }
 
+    /* Always unmask the optional set of signals so that we can support
```

Comment:
I thought one of the benefits of vDSO was that we no longer needed to rely on exception handlers; handlers are susceptible to order of registration and OE's handler cannot always  be the first.

My expectation was that in the vDSO case, we will disable registering the handler.

Reply:
When vDSO is used, the kernel will only filter out those hardware-based exceptions. The software-based signals (which we used to simulate enclave interrupt) won't be affected.

Reply:
In that case, is there any benefit to using vdso? We still have to register a handler.

Reply:
It's more relevant to the page faults when we support EDMM; the vDSO ensures the page faults will not be out-of-order so the we can safely use EDMM features inside the page fault handler.

Reply:
Ah Ok. In that case, how can the enclave be sure that the host is indeed using vDSO? Or does it not matter for the enclave whether vDSO is used to handle page faults or not?

Reply:
For enclave, whether vDSO is used or not does not matter because it's entirely the host-side implementation. For example, a malicious host can always force an out-of-order page fault against the enclave. This is more like a guarantee only when the host/kernel is trusted.

Reply:
> For enclave, whether vDSO is used or not does not matter because it's entirely the host-side implementation

Ok. Makes sense. 

> the vDSO ensures the page faults will not be out-of-order so the we can safely use EDMM features inside the page fault handler.

Does `safely use` refer to security or is it just performance?

> This is more like a guarantee only when the host/kernel is trusted.

Does this mean that EDMM necessarily requires trusting the host.

Note: I understand that these questions are inherent to EDMM/vDSO and not specific to this PR. I'm nonetheless interesting in updating my understanding.

Reply:
Let me add @haitaohuang who should provide more insights.

Based on https://github.com/haitaohuang/openenclave/blob/driver-intf/docs/DesignDocs/SGX2-EDMM-driver-interface.md#exception-handling, it only mentions

> The direct EENTER method requires signal handling in runtime library which is known to be challenging in Linux environment. Therefore, the vDSO interface is preferred and assumed in following discussion.

Reply:
Thanks for the explanation @haitaohuang. In this PR, we are still having to register signal handler when vDSO is used. Is that expected?

Reply:
Yes, My understanding OE needs support  signals for other purposes, like emulating  thread interruption